### ISSUE TITLE
[Rust] Implement three new Concept Exercises: basic-{fn, types, option}

### ISSUE THREAD

**Issue by workingjubilee** [exercism/v3#1022]
This issue breaks with the convention of a single issue per exercise because of the interrelatedness of these three exercises and the problem they intend to solve. They still need three patches, but it will likely benefit from a single issue for discussion as they rapidly build on each other and have a shared higher-level goal: overcoming some of the bigger newbie (and mentoring) Rust pain points, which are learning (or teaching) about functions, types, and unwrapping, as doing anything else with Rust depends on these.

Exercism is targeted at programmers of some experience but familiarity with the particular mix of C++ and OCaml idioms that Rust uses is not common, and even less so for students from further-out PLs like Python or Ruby. If Exercism is to successfully introduce Rust to others, it must quickly introduce the skillset that allows one to learn the rest of the language.

## Goal

These are intended to be model Rust exercises that show what is expected later on.
They are intended to teach the student enough to interface with Exercism.
They are intended to teach the student enough to interface with Rust's core APIs, so later exercises can assume a basic grasp of Rust usage and the student can go on learning how to interact with the rest of the Rust language.

The current intended divide is:
- `basic-fn`: teaches about type signatures and working with them, including some easy type casts. May use the pattern of [Rustlings]: provide "broken" code and request the reader to fix it after running `rustc` or `cargo check`, since main focus is on reading.
- `basic-types`: teaches about type annotations and variable declarations. Uses `const` to force learning of annotations, in an exception to the usual pattern.
- `basic-option`: teaches about basic usage of Option<T> and unwrapping to do more complex type conversions and working with uncertainty. The typical `unimplemented!("The exercise.")` stub should become familiar.

## Learning objectives

Students who complete these exercises should be able to...
- add comments for readability
- read [type signatures](https://en.wikipedia.org/wiki/Function_prototype)
- understand some basic consequences of type signatures, and thus
1. recognize some common primitive and a _few_ complex types
2. understand that type inference occurs
3. annotate basic types if inference fails
- declare variables using `let`
- declare variables using `const` (so as to force learning type annotation)
- perform some basic type casts
- use some basic conditional flow, in particular to unwrap Option types, requiring:
1. handling `None`
2. extracting `value` from `Some(value)`
3. using `if` or `match` to do something based on that
- use `use` to access namespaces
- annotate or modify code when `rustc` asks for it
- write idiomatic functions
1. use implicit return
2. use semicolons as separators instead of terminators

That's a lot, but that's why it's multiple exercises.

## Out of scope

Student **should not** need to _completely_ understand, but _if_ it is pedagogically expedient, **may** discover mentions of
- arithmetic
- the relationship of char, str, and String
- `std::collections`, methods, or slices
- `let mut` vs. immutability
- `if let` usage
- **very** basic usage of `match`
- a predefined macro or two like `format!`

Student **must not** need to understand
- Traits beyond `From` / `Into` / `TryFrom` / `TryInto`
- `fn main()` or handling IO
- ownership and lifetimes beyond `&str`
- More advanced `match` patterns
- More advanced control flow like loops and iterators
- General complex types, e.g. Structs, Enums, Tuples
- Result<T,E> and `panic!`
- `#[attributes]`, macros, or other compiler directives
- non-`std` crates

## Concepts

- `types`
- `fn`
- `option`
- `casting`
- `variables`
- `use`

## Prerequisites

None, hopefully.

## Resources to refer to

You can reference [The Book] as usual but in particular it is a good idea to read [Rust By Example] and [Rustlings].

### Hints

* Appropriate chapters of [The Book].
* The [standard library](https://doc.rust-lang.org/std/)

### After

* [Rust By Example]
* [Rustlings]

## Representer

No specific changes are expected to be required.

## Analyzer

No specific changes are expected to be required.

## Implementing

Please refer to https://github.com/exercism/v3/blob/master/languages/rust/reference/implementing-a-concept-exercise.md and implement the exercise in `/languages/rust/exercises/concept`.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

- [The Book]: https://doc.rust-lang.org/book/
- [Rust By Example]: https://doc.rust-lang.org/rust-by-example
- [Rustlings]: https://github.com/rust-lang/rustlings

**Comment by workingjubilee** [comment]
Some remarks after initial thought:

- I do not actually like the name "basic" for these, it has the same scent as using "simple" or "just" inappropriately or unnecessarily. "basic" can also mean "foundational" (intended), but I noticed my thoughts sorting everything into "basic" vs. "advanced" for no good reason. I switched to "intro" for the slugs because the temporal context (i.e. you are learning this first, and there will be more) is more clear to me.
- introducing `HashMap<K, V>` or `HashSet<T>` with `Option<T>` might be a great pairing for `intro-option`, as it also introduces `use`, and it's a simple binary question, reducing the complexity in the Some(value) to handle. This would suggest deferring [`From`], which would be okay as they would still end with the prerequisite skills to use `From`.

Today I learned:
- `fn` is a primitive type in Rust
- The [Rust prelude] is a Lot.
- The [identity function]
- The difference between [type conversion, coercion, and casting]
- [Traits are literally function signatures]

[Rust prelude]: https://doc.rust-lang.org/std/prelude/index.html
[identity function]: https://doc.rust-lang.org/std/convert/fn.identity.html
[type conversion, coercion, and casting]: https://en.wikipedia.org/wiki/Type_conversion
[Traits are literally function signatures]: https://doc.rust-lang.org/src/core/convert/mod.rs.html#381-385
[`From`]: https://doc.rust-lang.org/std/convert/trait.From.html

**Comment by seanchen1991** [comment]
Why `HashMap` or `HashSet` instead of `Vec` methods like `pop`, `first`, `last`, etc. ? I figure vectors should probably be introduced before `HashMap`s and `HashSet`s, and people will generally be more comfortable/familiar with the broad strokes of what vectors do.

**Comment by workingjubilee** [comment]
1) `Vec` is more common but would introduce reasoning about slices, and `array` vs `Vec`, et cetera. I would slightly prefer `array` and `Vec` are introduced simultaneously in an exercise so they could be discussed and differentiated. `HashMap` abstracts more completely over its backing store, thus letting us pretend it does not exist for the purpose of this exercise.
2) Because it permits simultaneous introduction of `use std::collections::` (`Vec` is in the prelude and should not be `use`d in most cases) to introduce Rust's namespacing syntax, while not making a big deal of it.

I am not married to a design for the exercise yet, of course: it's not even implemented.

**Comment by seanchen1991** [comment]
Agreed on your points about `Vec`, using `HashMap` instead would open up a less hairy can of worms.

**Comment by efx** [comment]
> If Exercism is to successfully introduce Rust to others, it must quickly introduce the skillset that allows one to learn the rest of the language.

Wonderful summary! This helps me frame our goals well as we draft the early exercises.

> use semicolons as separators instead of terminators

Curious, what are terminators?

**Comment by workingjubilee** [comment]
So, I first heard this explanation by someone who is not a Rust programmer, and it's not quite the same terminology the [Rust reference] uses. But to explain, I used "terminator" in the sense of a programming language's syntax: in C a semicolon terminates (ends) a simple `expression statement`. Consider this function in C:

```c
int do_math() {
  5 + 3;
}
```

Now consider this function in Rust:

```rust
fn do_math() -> i32 {
  5 + 3
}
```

Both of these are valid-ish even though the Rust one returns 8 and the C one "does nothing". Actually, the C one thinks it returns 0 according to compiling and printing the result with `gcc`, because C discards the value of an expression after computation. It requires an explicit directive in many programming languages to make a function return a value.

In Rust, the semicolon interrupts the implicit return behavior, separating expressions. This is a _logical_ alteration rather than a merely syntactic one. If we introduce a semicolon at the end of `do_math`'s function body, so it is `5 + 3;`, it causes the last thing there to be an _empty_ expression in effect, the same as the closure `let x = | | ;` which `dbg!` reports is `x() = ()`, returning the Unit type, and so with `do_math` that function fails to typecheck and `rustc` refuses to compile your code.

[Rust reference]: https://doc.rust-lang.org/reference/statements-and-expressions.html
